File: module_core/lib/shared_data/models/product_model.dart
import 'package:json_annotation/json_annotation.dart';
import '../../shared_domain/shared_entities/product.dart';
import 'timestamp_converter.dart';

part 'product_model.g.dart';

@JsonSerializable()
@TimestampConverter()
class ProductModel extends Product {
  const ProductModel({
    required super.uid,
    required super.gambar,
    @JsonKey(name: 'gambar_paths') required super.gambarPaths,
    @JsonKey(name: 'nama_brand') required super.namaBrand,
    required super.harga,
    required super.deskripsi,
    required super.detail,
    required super.sizes,
    @JsonKey(name: 'total_stok') required super.totalStok,
    required super.diskon,
    @JsonKey(name: 'harga_diskon') required super.hargaDiskon,
    @JsonKey(name: 'created_at') required super.createdAt,
    @JsonKey(name: 'updated_at') required super.updatedAt,
  });

  factory ProductModel.fromJson(Map<String, dynamic> json) => _$ProductModelFromJson(json);
  Map<String, dynamic> toJson() => _$ProductModelToJson(this);
}

File: functions/src/models/product.model.ts
export type SizeKey = "onesize" | "xs" | "s" | "m" | "l" | "xl" | "xxl";

export interface Product {
  uid: string;
  gambar: string[];
  gambar_paths: string[];
  nama_brand: string;
  harga: number;
  deskripsi: string;
  detail: string;
  sizes: Record<SizeKey, number>;
  total_stok: number;
  diskon: number;
  harga_diskon: number;
  created_at: FirebaseFirestore.Timestamp;
  updated_at: FirebaseFirestore.Timestamp;
}

export interface CreateProductInput {
  gambar_base64: string[];
  nama_brand: string;
  harga: number;
  deskripsi: string;
  detail: string;
  sizes: Record<string, number>;
  diskon?: number;
}

export interface UpdateProductInput {
  uid: string;
  gambar_base64?: string[];
  nama_brand?: string;
  harga?: number;
  deskripsi?: string;
  detail?: string;
  sizes?: Record<string, number>;
  diskon?: number;
}

export interface UpdateStockInput {
  uid: string;
  /** Map of size key to stock change (positive to add, negative to subtract) */
  stock_changes: Record<string, number>;
}

export interface DeleteProductInput {
  uid: string;
}
